p2p/net/swarm/dial_worker.go

```diff
@@ -268,46 +306,151 @@ func (w *dialWorker) dispatchError(ad *addrDial, err error) {
 	// this is necessary to support active listen scenarios, where a new dial comes in while
 	// another dial is in progress, and needs to do a direct connection without inhibitions from
 	// dial backoff.
-	// it is also necessary to preserve consisent behaviour with the old dialer -- TestDialBackoff
-	// regresses without this.
 	if err == ErrDialBackoff {
 		delete(w.pending, ad.addr)
 	}
 }
 
-// ranks addresses in descending order of preference for dialing, with the following rules:
-// NonRelay > Relay
-// NonWS > WS
-// Private > Public
-// UDP > TCP
-func (w *dialWorker) rankAddrs(addrs []ma.Multiaddr) []ma.Multiaddr {
-	addrTier := func(a ma.Multiaddr) (tier int) {
-		if isRelayAddr(a) {
-			tier |= 0b1000
+// rankAddrs ranks addresses for dialing. if it's a simConnect request we
+// dial all addresses immediately without any delay
+func (w *dialWorker) rankAddrs(addrs []ma.Multiaddr, isSimConnect bool) []network.AddrDelay {
+	if isSimConnect {
+		return noDelayRanker(addrs)
+	}
+	return w.s.dialRanker(addrs)
+}
+
+// dialQueue is a priority queue used to schedule dials
+type dialQueue struct {
+	// q is the queue maintained as a heap
+	q []network.AddrDelay
+	// pos is the reverse map from address to its position in q
+	// the reverse map is required to provide efficient updates
+	pos map[ma.Multiaddr]int
+}
+
+func newDialQueue() *dialQueue {
+	return &dialQueue{pos: make(map[ma.Multiaddr]int)}
+}
+
+// add adds adelay to the queue. if another elements exists in the queue with
+// the same address, it replaces that element.
+func (dq *dialQueue) add(adelay network.AddrDelay) {
```

Comment:
I agree with your convention.

I followed this because the convention you mentioned was not followed in some places like https://github.com/libp2p/go-libp2p/blob/master/p2p/net/swarm/dial_sync.go#L44
and I mistakenly assumed that the opposite was the convention. 

I'll fix this.

Reply:
fixed it. methods used outside and exported.